File: strategy/signal/trix_cross.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    StaticParameter,
)
from strategy.signal.base import Signal, SignalType


@dataclass(frozen=True)
class TRIXCrossSignal(Signal):
    type: SignalType = SignalType.TrixCross
    period: Parameter = StaticParameter(18.0)
    signal_period: Parameter = StaticParameter(9.0)

File: ta_lib/strategies/signal/src/trix_cross.rs
use base::prelude::*;
use core::prelude::*;
use momentum::trix;

pub struct TRIXCrossSignal {
    period: usize,
    signal_period: usize,
}

impl TRIXCrossSignal {
    pub fn new(period: f32, signal_period: f32) -> Self {
        Self {
            period: period as usize,
            signal_period: signal_period as usize,
        }
    }
}

impl Signal for TRIXCrossSignal {
    fn lookback(&self) -> usize {
        std::cmp::max(self.period, self.signal_period)
    }

    fn generate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let trix = trix(&data.close, self.period);
        let signal_line = trix.ma(self.signal_period);

        (
            trix.cross_over(&signal_line),
            trix.cross_under(&signal_line),
        )
    }
}
